### PR TITLE
Fixed #29229 -- Fixed column mismatch crash when combining two annotated values_list() querysets with union(), difference(), or intersection().

### PR DESCRIPTION
**Pull request by AmrAnwar**
https://code.djangoproject.com/ticket/29229

### CODE REVIEW

**Comment by AmrAnwar** [comment]
@charettes  ok course, I didn't see that it need a test from the site here: https://code.djangoproject.com/ticket/29229 
But I'll try now

**Comment by sauravjaiswalsj** [comment]
It is fixed i guess so.

**Comment by AmrAnwar** [comment]
@charettes  hope you check it now

**Comment by timgraham** [comment]
Can the test check the results of the query rather than internal implementation details like `.query.combined_queries[1].annotations`?

Incidentally, is there a way to write your query without `QuerySet.extra()`? That's likely to be deprecated in the future once we can obsolete it.

**Comment by AmrAnwar** [comment]
@timgraham  sorry for the delay it's my first time

File: tests/queries/test_qs_combinators.py [hunk 1]
```diff
@@ -119,6 +119,17 @@ def test_union_with_values(self):
         reserved_name = qs1.union(qs1).values_list('name', 'order', 'id').get()
         self.assertEqual(reserved_name[:2], ('a', 2))
 
+    def test_union_with_two_annotated_values_list(self):
+        qs1 = Number.objects.filter(num=1).annotate(
+            count=Value(0, IntegerField()),
+        ).values_list('id', 'count')
+        qs2 = Number.objects.filter(num=0).values('num').annotate(
+            id=F('pk'),
```
Comment (timgraham) [inline tests/queries/test_qs_combinators.py]: Relying on pk might be problematic since we shouldn't assume the values that the database might assign.

Comment (felixxm) [reply tests/queries/test_qs_combinators.py]: Agreed. Maybe:
```python
def test_union_with_two_annotated_values_list(self):
    qs1 = Number.objects.filter(num=1).annotate(
        count=Value(0, IntegerField()),
    ).values_list('num', 'count')
    qs2 = Number.objects.filter(num=2).values('pk').annotate(
        count=F('num'),
        num=Value(1, IntegerField()),
    ).values_list('num', 'count')
    self.assertCountEqual(qs1.union(qs2), [(1, 0), (2, 1)])
```

File: django/db/models/sql/compiler.py [hunk 2]
```diff
@@ -408,7 +408,8 @@ def get_combinator_sql(self, combinator, all):
                 # If the columns list is limited, then all combined queries
                 # must have the same columns list. Set the selects defined on
                 # the query on all combined queries, if not already set.
-                if not compiler.query.values_select and self.query.values_select:
+                if (not(compiler.query.values_select or compiler.query.annotations) and
+                        self.query.values_select):
```
[the remark this answers is not in the document]
Comment (AmrAnwar) [reply django/db/models/sql/compiler.py]: yes, got it thanks for your review

[the remark this answers is not in the document]
Comment (AmrAnwar) [reply django/db/models/sql/compiler.py]: I was going to push the code in the test function but your one better
```python
        qs1 = Number.objects.filter(num=1).annotate(
            count=Value(0, IntegerField())
        ).values('count', num2=F('num')).order_by()
        qs2 = Number.objects.filter(num=0).values('num').annotate(
            num2=F("num")).annotate(
            count=Value(0, IntegerField()),
        ).values_list('count', "num2")
        qs1_union_qs2 = qs2.union(qs1).order_by('num2')
        self.assertEqual(list(qs1_union_qs2), [(0, 0), (1, 0)])
```

**Comment by timgraham** [comment]
A couple builds failed, is the test non-deterministic?

**Comment by felixxm** [comment]
It should be deterministic, because the query is really simple:
```sql
SELECT "queries_number"."num", 0 AS "count"
FROM "queries_number"
WHERE "queries_number"."num" = 1
UNION 
SELECT "queries_number"."num" AS "count", 1 AS "num"
FROM "queries_number"
WHERE "queries_number"."num" = 2
```
The test failed only python 3.5, so maybe it revealed an unrelated bug  :thinking:. I will take a closer look at this.

**Comment by felixxm** [comment]
@timgraham Splitting annotations in the `qs2` fixed tests failure. `QuerySet.annotate()` doesn't preserve ordering of `kwargs` in python 3.5.

**Pull request merged**